source/mir/random/engine/package.d

```diff
@@ -151,3 +143,378 @@ unittest
+        if (res == 0)
+        {
+            if (fdURandom.ferror)
+                return -1;
+
+            if (fdURandom.feof)
+                return -1;
+        }
+        return res;
+    }
+}
+
+version(Windows)
+{
+    // the wincrypt headers in druntime are broken for x64!
+    private alias ULONG_PTR = size_t; // uint in druntime
+    private alias BOOL = bool;
+    private alias DWORD = size_t; // uint in druntime
+    private alias LPCWSTR = wchar*;
+    private alias PBYTE = ubyte*;
+    private alias HCRYPTPROV = ULONG_PTR;
+    private alias LPCSTR = const(char)*;
+
+    private extern(Windows) BOOL CryptGenRandom(HCRYPTPROV, DWORD, PBYTE) @nogc @safe nothrow;
+    private extern(Windows) BOOL CryptAcquireContextA(HCRYPTPROV*, LPCSTR, LPCSTR, DWORD, DWORD) @nogc nothrow;
+    private extern(Windows) BOOL CryptAcquireContextW(HCRYPTPROV*, LPCWSTR, LPCWSTR, DWORD, DWORD) @nogc nothrow;
+    private extern(Windows) BOOL CryptReleaseContext(HCRYPTPROV, ULONG_PTR) @nogc nothrow;
+
+    private __gshared ULONG_PTR hProvider;
+
+    private auto initGetRandom() @nogc @trusted nothrow
+    {
+        import core.sys.windows.winbase : GetLastError;
+        import core.sys.windows.winerror : NTE_BAD_KEYSET;
+        import core.sys.windows.wincrypt : PROV_RSA_FULL, CRYPT_NEWKEYSET, CRYPT_VERIFYCONTEXT, CRYPT_SILENT;
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/aa379886(v=vs.85).aspx
+        // For performance reasons, we recommend that you set the pszContainer
+        // parameter to NULL and the dwFlags parameter to CRYPT_VERIFYCONTEXT
+        // in all situations where you do not require a persisted key.
+        // CRYPT_SILENT is intended for use with applications for which the UI cannot be displayed by the CSP.
+        if (!CryptAcquireContextW(&hProvider, null, null, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
+        {
+            if (GetLastError() == NTE_BAD_KEYSET)
+            {
+                // Attempt to create default container
+                if (!CryptAcquireContextA(&hProvider, null, null, PROV_RSA_FULL, CRYPT_NEWKEYSET | CRYPT_SILENT))
+                    return 1;
+            }
+            else
+            {
+                return 1;
+            }
+        }
+
+        return 0;
+    }
+
+    extern(C) shared static ~this()
+    {
+        if (hProvider > 0)
+            CryptReleaseContext(hProvider, 0);
+    }
+}
+
+/**
+Fills a buffer with random data.
+If not enough entropy has been gathered, it will block.
+
+Params:
+    ptr = pointer to the buffer to fill
+    len = length of the buffer (in bytes)
+
+Returns:
+    A non-zero integer if an error occurred.
+*/
+extern(C) ptrdiff_t mir_random_genRandomBlocking(void* ptr , size_t len) @nogc @trusted nothrow
+{
+    version(Windows)
+    {
+        if (hProvider == 0)
+            if (initGetRandom != 0)
+                return -1;
+
+        while(!CryptGenRandom(hProvider, len, cast(PBYTE) ptr)) {}
```

Comment:
For Windows all [error codes](https://msdn.microsoft.com/en-us/library/windows/desktop/aa379942(v=vs.85).aspx) of `CryptGenRandom` seem unrecoverable, so the loop here doesn't seem very useful. I think we should directly yield the error here.